File: src/java/magmac/api/result/Ok.java
package magmac.api.result;

import magmac.api.Tuple2;

import java.util.function.Function;
import java.util.function.Supplier;

public record Ok<T, X>(T value) implements Result<T, X> {
    @Override
    public <R> Result<R, X> mapValue(Function<T, R> mapper) {
        return new Ok<>(mapper.apply(this.value));
    }

    @Override
    public <R> Result<Tuple2<T, R>, X> and(Supplier<Result<R, X>> supplier) {
        return supplier.get().mapValue(otherValue -> new Tuple2<>(this.value, otherValue));
    }

    @Override
    public <R> R match(Function<T, R> whenOk, Function<X, R> whenErr) {
        return whenOk.apply(this.value);
    }

    @Override
    public <R> Result<R, X> flatMapValue(Function<T, Result<R, X>> mapper) {
        return mapper.apply(this.value);
    }

    @Override
    public <R> Result<T, R> mapErr(Function<X, R> mapper) {
        return new Ok<>(this.value);
    }
}

File: src/web/magmac/api/result/Ok.ts
import { Tuple2 } from "../../../magmac/api/Tuple2";
import { Function } from "../../../java/util/function/Function";
import { Supplier } from "../../../java/util/function/Supplier";
export class Ok<T,  X> {
	mapValue(mapper : Function<T, R>) : Result<R, X> {return 0;;}
	and(supplier : Supplier<Result<R, X>>) : Result<Tuple2<T, R>, X> {return 0;;}
	match(whenOk : Function<T, R>, whenErr : Function<X, R>) : R {return 0;;}
	flatMapValue(mapper : Function<T, Result<R, X>>) : Result<R, X> {return 0;;}
	mapErr(mapper : Function<X, R>) : Result<T, R> {return 0;;}
}
